File: lib/logging/f_logger_base.dart
// Copyright 2023, the Freemework.ORG project authors. Please see the AUTHORS
// file for details. All rights reserved. Use of this source code is governed
// by a BSD-style license that can be found in the LICENSE file.

import "dart:collection" show UnmodifiableMapView;

import "package:meta/meta.dart";

import "../exception/f_exception.dart" show FException;
import "../exception/f_exception_argument.dart" show FExceptionArgument;
import "../execution_context/f_execution_context.dart" show FExecutionContext;

import "f_logger.dart" show FLogger, FLoggerMessageFactory;
import "f_logger_labels.dart" show FLoggerLabels;
import "f_logger_labels_execution_context.dart"
    show FLoggerLabelsExecutionContext;
import "f_logger_level.dart" show FLoggerLevel;

abstract class FLoggerBase extends FLogger {
  @override
  bool get isTraceEnabled {
    return this.isLevelEnabled(FLoggerLevel.trace);
  }

  @override
  bool get isDebugEnabled {
    return this.isLevelEnabled(FLoggerLevel.debug);
  }

  @override
  bool get isInfoEnabled {
    return this.isLevelEnabled(FLoggerLevel.info);
  }

  @override
  bool get isWarnEnabled {
    return this.isLevelEnabled(FLoggerLevel.warn);
  }

  @override
  bool get isErrorEnabled {
    return this.isLevelEnabled(FLoggerLevel.error);
  }

  @override
  bool get isFatalEnabled {
    return this.isLevelEnabled(FLoggerLevel.fatal);
  }

  @override
  String? get name {
    return this._name;
  }

  @override
  void trace(
    dynamic labelsOrExecutionContext,
    dynamic messageOrMessageFactory, [
    FException? ex,
  ]) {
    if (!this.isTraceEnabled) {
      return;
    }
    final FLoggerLabels loggerLabels =
        FLoggerBase._resolveLoggerLabels(labelsOrExecutionContext);
    final String message = FLoggerBase._resolveMessage(messageOrMessageFactory);
    this.log(FLoggerLevel.trace, loggerLabels, message, ex);
  }

  @override
  void debug(
    dynamic labelsOrExecutionContext,
    dynamic messageOrMessageFactory, [
    FException? ex,
  ]) {
    if (!this.isDebugEnabled) {
      return;
    }
    final FLoggerLabels loggerLabels =
        FLoggerBase._resolveLoggerLabels(labelsOrExecutionContext);
    final String message = FLoggerBase._resolveMessage(messageOrMessageFactory);
    this.log(FLoggerLevel.debug, loggerLabels, message, ex);
  }

  @override
  void info(
    dynamic labelsOrExecutionContext,
    dynamic messageOrMessageFactory,
  ) {
    if (!this.isInfoEnabled) {
      return;
    }
    final FLoggerLabels loggerLabels =
        FLoggerBase._resolveLoggerLabels(labelsOrExecutionContext);
    final String message = FLoggerBase._resolveMessage(messageOrMessageFactory);
    this.log(FLoggerLevel.info, loggerLabels, message);
  }

  @override
  void warn(
    dynamic labelsOrExecutionContext,
    dynamic messageOrMessageFactory,
  ) {
    if (!this.isWarnEnabled) {
      return;
    }
    final FLoggerLabels loggerLabels =
        FLoggerBase._resolveLoggerLabels(labelsOrExecutionContext);
    final String message = FLoggerBase._resolveMessage(messageOrMessageFactory);
    this.log(FLoggerLevel.warn, loggerLabels, message);
  }

  @override
  void error(
    dynamic labelsOrExecutionContext,
    dynamic messageOrMessageFactory,
  ) {
    if (!this.isErrorEnabled) {
      return;
    }
    final FLoggerLabels loggerLabels =
        FLoggerBase._resolveLoggerLabels(labelsOrExecutionContext);
    final String message = FLoggerBase._resolveMessage(messageOrMessageFactory);
    this.log(FLoggerLevel.error, loggerLabels, message);
  }

  @override
  void fatal(
    dynamic labelsOrExecutionContext,
    dynamic messageOrMessageFactory,
  ) {
    if (!this.isFatalEnabled) {
      return;
    }
    final FLoggerLabels loggerLabels =
        FLoggerBase._resolveLoggerLabels(labelsOrExecutionContext);
    final String message = FLoggerBase._resolveMessage(messageOrMessageFactory);
    this.log(FLoggerLevel.fatal, loggerLabels, message);
  }

  @protected
  bool isLevelEnabled(FLoggerLevel level);

  ///
  /// Override this method to implement custom logger
  ///
  @protected
  void log(
    FLoggerLevel level,
    FLoggerLabels labels,
    String message, [
    FException? exception,
  ]);

  @protected
  FLoggerBase(String? loggerName) : this._name = loggerName;

  final String? _name;

  ///
  /// There two overloads
  /// ```dart
  /// FLoggerLabels _resolveLoggerLabels(FExecutionContext executionContext);
  /// FLoggerLabels _resolveLoggerLabels(FLoggerLabels labels);
  /// ```
  ///
  static FLoggerLabels _resolveLoggerLabels(
    dynamic labelsOrExecutionContext,
  ) {
    if (labelsOrExecutionContext == null) {
      return FLoggerBase._emptyLabels;
    } else if (labelsOrExecutionContext is FExecutionContext) {
      final executionElement =
          FLoggerLabelsExecutionContext.of(labelsOrExecutionContext);
      if (executionElement != null) {
        return UnmodifiableMapView<String, String>(
          {...executionElement.loggerLabels},
        );
      } else {
        // No any logger properties on excecution context chain
        return FLoggerBase._emptyLabels;
      }
    } else if (labelsOrExecutionContext is FLoggerLabels) {
      return labelsOrExecutionContext;
    } else {
      throw FExceptionArgument(
        "Unexpected overloading. Unsupported type '${labelsOrExecutionContext.runtimeType}'",
        "labelsOrExecutionContext",
      );
    }
  }

  ///
  /// There two overloads
  /// ```dart
  /// String _resolveMessage(String message) => message;
  /// String _resolveMessage(FLoggerMessageFactory messageFactory) => messageFactory()
  /// ```
  ///
  static String _resolveMessage(dynamic messageOrMessageFactory) {
    if (messageOrMessageFactory is String) {
      return messageOrMessageFactory;
    } else if (messageOrMessageFactory is FLoggerMessageFactory) {
      return messageOrMessageFactory();
    } else {
      throw FExceptionArgument(
        "Unexpected overloading. Unsupported type '${messageOrMessageFactory.runtimeType}'. Expected String or FLoggerMessageFactory.",
        "messageOrMessageFactory",
      );
    }
  }

  static final UnmodifiableMapView<String, String> _emptyLabels =
      UnmodifiableMapView({});
}

File: src/logging/f_logger_base.ts
import { FException } from "../exception/f_exception";
import { FExceptionArgument } from "../exception/f_exception_argument";
import { FExecutionContext } from "../execution_context/f_execution_context";

import { FLogger, FLoggerMessageFactory } from "./f_logger";
import { FLoggerLabels } from "./f_logger_labels";
import { FLoggerLabelsExecutionContext } from "./f_logger_labels_execution_context";
import { FLoggerLevel } from "./f_logger_level";

export abstract class FLoggerBase extends FLogger {
	public get isTraceEnabled(): boolean { return this.isLevelEnabled(FLoggerLevel.TRACE); }
	public get isDebugEnabled(): boolean { return this.isLevelEnabled(FLoggerLevel.DEBUG); }
	public get isInfoEnabled(): boolean { return this.isLevelEnabled(FLoggerLevel.INFO); }
	public get isWarnEnabled(): boolean { return this.isLevelEnabled(FLoggerLevel.WARN); }
	public get isErrorEnabled(): boolean { return this.isLevelEnabled(FLoggerLevel.ERROR); }
	public get isFatalEnabled(): boolean { return this.isLevelEnabled(FLoggerLevel.FATAL); }

	public get name(): string { return this._name; }

	public trace(
		variant: FExecutionContext | FLoggerLabels,
		messageOrMessageFactory: string | FLoggerMessageFactory,
		ex?: FException,
	): void {
		if (!this.isTraceEnabled) {
			return;
		}

		const loggerLabels: FLoggerLabels = FLoggerBase._resolveLoggerLabels(variant);
		const message: string = FLoggerBase._resolveMessage(messageOrMessageFactory);

		this.log(FLoggerLevel.TRACE, loggerLabels, message, ex);
	}

	public debug(
		variant: FExecutionContext | FLoggerLabels,
		messageOrMessageFactory: string | FLoggerMessageFactory,
		ex?: FException,
	): void {
		if (!this.isDebugEnabled) {
			return;
		}

		const loggerLabels: FLoggerLabels =
			FLoggerBase._resolveLoggerLabels(variant);
		const message: string = FLoggerBase._resolveMessage(messageOrMessageFactory);
		this.log(FLoggerLevel.DEBUG, loggerLabels, message, ex);
	}

	public info(
		variant: FExecutionContext | FLoggerLabels,
		messageOrMessageFactory: string | FLoggerMessageFactory,
	): void {
		if (!this.isInfoEnabled) {
			return;
		}

		const loggerLabels: FLoggerLabels =
			FLoggerBase._resolveLoggerLabels(variant);
		const message: string = FLoggerBase._resolveMessage(messageOrMessageFactory);
		this.log(FLoggerLevel.INFO, loggerLabels, message);
	}

	public warn(
		variant: FExecutionContext | FLoggerLabels,
		messageOrMessageFactory: string | FLoggerMessageFactory,
	): void {
		if (!this.isWarnEnabled) {
			return;
		}

		const loggerLabels: FLoggerLabels =
			FLoggerBase._resolveLoggerLabels(variant);
		const message: string = FLoggerBase._resolveMessage(messageOrMessageFactory);
		this.log(FLoggerLevel.WARN, loggerLabels, message);
	}

	public error(
		variant: FExecutionContext | FLoggerLabels,
		messageOrMessageFactory: string | FLoggerMessageFactory,
	): void {
		if (!this.isErrorEnabled) {
			return;
		}

		const loggerLabels: FLoggerLabels =
			FLoggerBase._resolveLoggerLabels(variant);
		const message: string = FLoggerBase._resolveMessage(messageOrMessageFactory);
		this.log(FLoggerLevel.ERROR, loggerLabels, message);
	}

	public fatal(
		variant: FExecutionContext | FLoggerLabels,
		messageOrMessageFactory: string | FLoggerMessageFactory,
	): void {
		if (!this.isFatalEnabled) {
			return;
		}

		const loggerLabels: FLoggerLabels =
			FLoggerBase._resolveLoggerLabels(variant);
		const message: string = FLoggerBase._resolveMessage(messageOrMessageFactory);
		this.log(FLoggerLevel.FATAL, loggerLabels, message);
	}

	protected constructor(loggerName: string) {
		super();

		this._name = loggerName;
	}

	protected abstract isLevelEnabled(level: FLoggerLevel): boolean;

	///
	/// Override this method to implement custom logger
	///
	protected abstract log(
		level: FLoggerLevel,
		labels: FLoggerLabels,
		message: string,
		exception?: FException,
	): void;

	private readonly _name: string;

	private static _resolveLoggerLabels(
		variant: FExecutionContext | FLoggerLabels,
	): FLoggerLabels {
		if (variant === null || variant === undefined) {
			// Sometime users pass undefined/null value.
			// It is contract violation, but not a reason to crash in logger
			return FLoggerBase._emptyLabels; 
		} else if (variant instanceof FExecutionContext) {
			const executionElement = FLoggerLabelsExecutionContext
				.of(variant);
			if (executionElement !== null) {
				return Object.freeze({ ...executionElement.loggerLabels });
			} else {
				return FLoggerBase._emptyLabels; // No any logger properties on excecution context chain
			}
		} else {
			return variant;
		}
	}

	private static _resolveMessage(messageOrMessageFactory: string | FLoggerMessageFactory): string {
		if (typeof messageOrMessageFactory === "function") {
			try {
				messageOrMessageFactory = messageOrMessageFactory();
				if (typeof messageOrMessageFactory !== "string") {
					messageOrMessageFactory = `FLoggerMessageFactory contract violation detected: Non-string result. Force stringify message: ${messageOrMessageFactory}`;
				}
			} catch (e) {
				messageOrMessageFactory = "FLoggerMessageFactory contract violation detected: Exception was raised."
			}
		} else {
			if (typeof messageOrMessageFactory !== "string") {
				messageOrMessageFactory = `FLogger contract violation detected: Non-string message passed. Force stringify message: ${messageOrMessageFactory}`;
			}
		}

		return messageOrMessageFactory;
	}

	private static readonly _emptyLabels: FLoggerLabels = Object.freeze({});
}
